File: Examples/GroupDocs.Metadata.Examples.VBasic/Utilities/MyExtension.vb
Imports GroupDocs.Metadata
Imports GroupDocs.Metadata.Tools
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
'ExStart:DocumentTypeDetectorDirectoryExtension
Namespace System.IO
    Public Module MyExtesion
        <Runtime.CompilerServices.Extension()> _
        Public Function GetFiles(directoryInfo As Global.System.IO.DirectoryInfo, documentType As DocumentType) As Global.System.IO.FileInfo()
            Dim files As Global.System.IO.FileInfo() = directoryInfo.GetFiles()

            ' return empty array if directory is empty
            If files.Length = 0 Then
                Return New Global.System.IO.FileInfo(-1) {}
            End If

            Dim result As New List(Of Global.System.IO.FileInfo)()

            For Each fileInfo As Global.System.IO.FileInfo In files
                Using fileFormatChecker As New FileFormatChecker(fileInfo.FullName)
                    If fileFormatChecker.VerifyFormat(documentType) Then
                        result.Add(fileInfo)
                    End If
                End Using
            Next

            Return result.ToArray()
        End Function
    End Module
End Namespace
'ExEnd:DocumentTypeDetectorDirectoryExtension

File: Examples/GroupDocs.Metadata.Examples.CSharp/Utilities/MyExtension.cs
using GroupDocs.Metadata;
using GroupDocs.Metadata.Tools;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

//ExStart:DocumentTypeDetectorDirectoryExtension
namespace System.IO
{
    public static class MyExtension
    {
        /// <summary>
        /// Gets and returns files of a specific document type
        /// </summary>
        /// <param name="directory">Directory Path</param>
        /// <param name="documentType">Document Type</param>
        /// <returns>File info array</returns>
        public static FileInfo[] GetFiles(this DirectoryInfo directoryInfo, DocumentType documentType)
        {
            FileInfo[] files = directoryInfo.GetFiles();

            // return empty array if directory is empty
            if (files.Length == 0)
            {
                return new FileInfo[0];
            }

            List<FileInfo> result = new List<FileInfo>();

            foreach (FileInfo fileInfo in files)
            {
                using (FileFormatChecker fileFormatChecker = new FileFormatChecker(fileInfo.FullName))
                {
                    if (fileFormatChecker.VerifyFormat(documentType))
                    {
                        result.Add(fileInfo);
                    }
                }
            }

            return result.ToArray();
        }
    }
}
//ExEnd:DocumentTypeDetectorDirectoryExtension
